### ISSUE TITLE
max request size limitation without buffer

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#36946]
*Title*: *One line description*

Now, all the the exist way to limit the max request body in the Envoy need to buffer the request body. But I won't whether make sense to add a featue to limit the request body without buffer body.

The request body will be proxied to upstream. But the the limitation is reached then the 413 will be returned and whole connection will be closed directly.

From the clients' view, the Envoy responded a 413, from the servers' view, Envoy cancelled the requests by close connection. This could limit the max request body size and need keep the body in the memory buffer. 

🤔

**Comment by wufanqqfsc** [comment]
It's reasonable to return 413 ,but why we need close the connection ? The max request body size reached for this request but not mean other requests via this connection will also over size right ?

**Comment by wbpcode** [comment]
> It's reasonable to return 413 ,but why we need close the connection ? The max request body size reached for this request but not mean other requests via this connection will also over size right ?

It depends on the protocol. HTTP1.1 will close the connection because the reponse will be complete before the response. For the HTTP2, the connection may be kept.

**Comment by wbpcode** [comment]
> To confirm the problem: we want to have a max request payload size while streaming the request to avoid buffering the entire request in memory at the proxy?

yeah.

**Comment by wufanqqfsc** [comment]
> > It's reasonable to return 413 ,but why we need close the connection ? The max request body size reached for this request but not mean other requests via this connection will also over size right ?
> 
> It depends on the protocol. HTTP1.1 will close the connection because the reponse will be complete before the response. For the HTTP2, the connection may be kept.

What about the http1.1 with keep-alive?

**Comment by wbpcode** [comment]
> > > It's reasonable to return 413 ,but why we need close the connection ? The max request body size reached for this request but not mean other requests via this connection will also over size right ?
> > 
> > 
> > It depends on the protocol. HTTP1.1 will close the connection because the reponse will be complete before the response. For the HTTP2, the connection may be kept.
> 
> What about the http1.1 with keep-alive?

HTTP1.1 has no frame or stream like HTTP2 on the top of the connection. So, if the response cannot end as normally, to avoid possible dirty data, the connection should be closed anyway.

**Comment by wufanqqfsc** [comment]
OK, the proposed solution  here is:
1.For downstream: envoy count and sum the received buffer size. And then compare with our target max body size. If over size then envoy sendLocalrepay to downstream with 413. 
2.For upstream you want to close the connection to avoid more dirty data transfer ed to upstream service right?  But anyway there is already some dirty data send to the upstream. So we want envoy to close the 
connection. And for the upstream service should handle the connection lost event and clean the buffer data.

I just wonder how this body limit was done by other proxy like nginx . Is it possible to have the same behavior ? So that it will much easier to do proxy replacement and no sence for upstream service.

**Comment by wbpcode** [comment]
> I just wonder how this body limit was done by other proxy like nginx . Is it possible to have the same behavior ? So that it will much easier to do proxy replacement and no sence for upstream service.

Nginx may buffer whole request, but Envoy will handle the request as a stream, it's a little different.

**Comment by ravenblackx** [comment]
Wouldn't this be fairly simple to implement as an HttpFilter? It would be pretty much a simplified version of the buffer filter, where instead of buffering it just counts, and does a localReply error if the size exceeds a configured limit (plus if the size is known from the headers, do the same rejection early).

And invoking a localReply would automatically resolve whatever is left over with the upstream in an appropriate way.

**Comment by wbpcode** [comment]
> Wouldn't this be fairly simple to implement as an HttpFilter? It would be pretty much a simplified version of the buffer filter, where instead of buffering it just counts, and does a localReply error if the size exceeds a configured limit (plus if the size is known from the headers, do the same rejection early).
> 
> And invoking a localReply would automatically resolve whatever is left over with the upstream in an appropriate way.

Yeah. I think so. But we didn't has a filter like this. (actually, we can also do it simply with lua script, but most normal users may didn't won't to handle the complexity of Lua script.

**Comment by Sreini** [comment]
it would be super nice if this was a feature of envoy
